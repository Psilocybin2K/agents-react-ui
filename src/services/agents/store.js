import React, { createContext, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
import { listAgents as svcList, createAgent as svcCreate, updateAgent as svcUpdate, deleteAgent as svcDelete } from './index';

const AgentsContext = createContext(null);

function sortByUpdatedAtDesc(items) {
  return items.slice().sort((a, b) => (b.updatedAt || 0) - (a.updatedAt || 0));
}

export function AgentsProvider({ children }) {
  const [agents, setAgents] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);
  const mountedRef = useRef(true);

  useEffect(() => {
    mountedRef.current = true;
    (async () => {
      try {
        const data = await svcList();
        if (mountedRef.current) setAgents(sortByUpdatedAtDesc(data));
      } catch (err) {
        if (mountedRef.current) setError(err);
      } finally {
        if (mountedRef.current) setIsLoading(false);
      }
    })();
    return () => { mountedRef.current = false; };
  }, []);

  const refresh = useCallback(async () => {
    const data = await svcList();
    if (mountedRef.current) setAgents(sortByUpdatedAtDesc(data));
  }, []);

  const createAgent = useCallback(async (data) => {
    const created = await svcCreate(data);
    if (mountedRef.current) setAgents(prev => sortByUpdatedAtDesc([...prev, created]));
    return created;
  }, []);

  const updateAgent = useCallback(async (id, updates) => {
    const updated = await svcUpdate(id, updates);
    if (mountedRef.current) setAgents(prev => {
      const idx = prev.findIndex(a => a.id === id);
      if (idx === -1) return prev;
      const next = prev.slice();
      next[idx] = updated;
      return sortByUpdatedAtDesc(next);
    });
    return updated;
  }, []);

  const deleteAgent = useCallback(async (id) => {
    await svcDelete(id);
    if (mountedRef.current) setAgents(prev => prev.filter(a => a.id !== id));
  }, []);

  const value = useMemo(() => ({
    agents,
    isLoading,
    error,
    refresh,
    createAgent,
    updateAgent,
    deleteAgent,
  }), [agents, isLoading, error, refresh, createAgent, updateAgent, deleteAgent]);

  return (
    <AgentsContext.Provider value={value}>
      {children}
    </AgentsContext.Provider>
  );
}

export function useAgents() {
  const ctx = useContext(AgentsContext);
  if (!ctx) {
    throw new Error('useAgents must be used within an AgentsProvider');
  }
  return ctx;
}


