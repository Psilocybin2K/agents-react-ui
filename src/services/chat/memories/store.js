import React, { createContext, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
import { listMemories as svcList, createMemory as svcCreate, updateMemory as svcUpdate, deleteMemory as svcDelete, generateMemoriesFromMessages as svcGenerate } from './index';

const ChatMemoriesContext = createContext(null);

function sortByUpdatedAtDesc(items) {
  return items.slice().sort((a, b) => (b.updatedAt || 0) - (a.updatedAt || 0));
}

export function ChatMemoriesProvider({ threadId, children }) {
  const [memories, setMemories] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);
  const mountedRef = useRef(true);
  const threadRef = useRef(threadId);

  useEffect(() => {
    threadRef.current = threadId;
  }, [threadId]);

  useEffect(() => {
    mountedRef.current = true;
    (async () => {
      try {
        setIsLoading(true);
        const data = await svcList(threadId);
        if (mountedRef.current) setMemories(sortByUpdatedAtDesc(data));
      } catch (err) {
        if (mountedRef.current) setError(err);
      } finally {
        if (mountedRef.current) setIsLoading(false);
      }
    })();
    return () => { mountedRef.current = false; };
  }, [threadId]);

  const refresh = useCallback(async () => {
    const data = await svcList(threadRef.current);
    if (mountedRef.current) setMemories(sortByUpdatedAtDesc(data));
  }, []);

  const createMemory = useCallback(async (data) => {
    const created = await svcCreate(threadRef.current, data);
    if (mountedRef.current) setMemories(prev => sortByUpdatedAtDesc([...prev, created]));
    return created;
  }, []);

  const updateMemory = useCallback(async (id, updates) => {
    const updated = await svcUpdate(threadRef.current, id, updates);
    if (mountedRef.current) setMemories(prev => {
      const idx = prev.findIndex(m => m.id === id);
      if (idx === -1) return prev;
      const next = prev.slice();
      next[idx] = updated;
      return sortByUpdatedAtDesc(next);
    });
    return updated;
  }, []);

  const deleteMemory = useCallback(async (id) => {
    await svcDelete(threadRef.current, id);
    if (mountedRef.current) setMemories(prev => prev.filter(m => m.id !== id));
  }, []);

  const generate = useCallback(async (messages, options) => {
    return svcGenerate(messages, options);
  }, []);

  const value = useMemo(() => ({
    threadId: threadRef.current,
    memories,
    isLoading,
    error,
    refresh,
    createMemory,
    updateMemory,
    deleteMemory,
    generate,
  }), [memories, isLoading, error, refresh, createMemory, updateMemory, deleteMemory, generate]);

  return (
    <ChatMemoriesContext.Provider value={value}>
      {children}
    </ChatMemoriesContext.Provider>
  );
}

export function useChatMemories() {
  const ctx = useContext(ChatMemoriesContext);
  if (!ctx) {
    throw new Error('useChatMemories must be used within a ChatMemoriesProvider');
  }
  return ctx;
}


