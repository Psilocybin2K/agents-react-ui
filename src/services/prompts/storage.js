// Lightweight localStorage adapter for Prompt records
// Storage shape: Array<Prompt> serialized as JSON under STORAGE_KEY
/**
 * @typedef {import('../../types').Prompt} Prompt
 */

const STORAGE_KEY = 'ado.prompts.v1';

function getLocalStorage() {
  if (typeof window === 'undefined') return null;
  try {
    return window.localStorage || null;
  } catch (_) {
    return null;
  }
}

function safeParse(json, fallback) {
  try {
    const value = JSON.parse(json);
    return Array.isArray(value) ? value : fallback;
  } catch (_) {
    return fallback;
  }
}

/**
 * Read all prompts from storage.
 * @returns {Prompt[]} prompts array (could be empty)
 */
export function getAll() {
  const ls = getLocalStorage();
  if (!ls) return [];
  const raw = ls.getItem(STORAGE_KEY);
  if (!raw) return [];
  return safeParse(raw, []);
}

/**
 * Overwrite storage with the provided prompts array.
 * @param {Prompt[]} prompts
 */
export function setAll(prompts) {
  const ls = getLocalStorage();
  if (!ls) return;
  ls.setItem(STORAGE_KEY, JSON.stringify(prompts || []));
}

/**
 * Clear all prompts from storage.
 */
export function clear() {
  const ls = getLocalStorage();
  if (!ls) return;
  ls.removeItem(STORAGE_KEY);
}

/**
 * Seed storage with provided prompts if currently empty.
 * @param {Prompt[]} seedPrompts
 * @returns {boolean} true if seeding occurred
 */
export function seedIfEmpty(seedPrompts) {
  const current = getAll();
  if (current.length > 0) return false;
  if (!Array.isArray(seedPrompts) || seedPrompts.length === 0) return false;
  setAll(seedPrompts);
  return true;
}


