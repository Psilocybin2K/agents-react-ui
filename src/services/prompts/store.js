import React, { createContext, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
import * as service from './index';

const PromptsContext = createContext(null);

export function PromptsProvider({ children }) {
  const mountedRef = useRef(false);
  const [prompts, setPrompts] = useState([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);

  function tempId() {
    return 'tmp_' + Math.random().toString(36).slice(2, 10) + '_' + Date.now().toString(36);
  }

  const refresh = useCallback(async () => {
    setLoading(true);
    setError(null);
    try {
      const list = await service.listPrompts();
      setPrompts(prev => {
        const fetched = Array.isArray(list) ? list : [];
        const optimistic = Array.isArray(prev) ? prev.filter(p => p.__optimistic) : [];
        const merged = [...optimistic, ...fetched];
        return merged;
      });
    } catch (err) {
      setError(err);
    } finally {
      setLoading(false);
    }
  }, []);

  useEffect(() => {
    if (mountedRef.current) return;
    mountedRef.current = true;
    refresh();
  }, [refresh]);

  const sortByUpdatedAtDesc = useCallback((items) => items.slice().sort((a, b) => (b.updatedAt || 0) - (a.updatedAt || 0)), []);

  const createPrompt = useCallback(async (data) => {
    // Optimistic insert
    const optimistic = {
      id: tempId(),
      name: data?.name || '',
      description: data?.description || '',
      content: data?.content || '',
      variables: Array.isArray(data?.variables) ? data.variables : [],
      tags: Array.isArray(data?.tags) ? data.tags : [],
      createdAt: Date.now(),
      updatedAt: Date.now(),
      __optimistic: true,
    };
    setPrompts(prev => [optimistic, ...prev]);
    try {
      const created = await service.createPrompt(data);
      setPrompts(prev => {
        const withoutOptimistic = (prev || []).filter(p => !p.__optimistic);
        return sortByUpdatedAtDesc([created, ...withoutOptimistic]);
      });
      return created;
    } catch (err) {
      setPrompts(prev => prev.filter(p => p.id !== optimistic.id));
      setError(err);
      throw err;
    }
  }, [tempId, sortByUpdatedAtDesc]);

  const updatePrompt = useCallback(async (id, updates) => {
    const prevList = prompts;
    // Optimistic update
    setPrompts(prev => sortByUpdatedAtDesc(prev.map(p => (p.id === id ? { ...p, ...updates, updatedAt: Date.now() } : p))));
    try {
      const updated = await service.updatePrompt(id, updates);
      setPrompts(prev => sortByUpdatedAtDesc(prev.map(p => (p.id === id ? updated : p))));
      return updated;
    } catch (err) {
      setPrompts(prevList);
      setError(err);
      throw err;
    }
  }, [prompts, sortByUpdatedAtDesc]);

  const deletePrompt = useCallback(async (id) => {
    const prevList = prompts;
    // Optimistic remove
    setPrompts(prev => prev.filter(p => p.id !== id));
    try {
      await service.deletePrompt(id);
    } catch (err) {
      setPrompts(prevList);
      setError(err);
      throw err;
    }
  }, [prompts]);

  const value = useMemo(() => ({
    prompts,
    loading,
    error,
    refresh,
    createPrompt,
    updatePrompt,
    deletePrompt,
  }), [prompts, loading, error, refresh, createPrompt, updatePrompt, deletePrompt]);

  return (
    <PromptsContext.Provider value={value}>
      {children}
    </PromptsContext.Provider>
  );
}

export function usePrompts() {
  const ctx = useContext(PromptsContext);
  if (!ctx) {
    throw new Error('usePrompts must be used within a PromptsProvider');
  }
  return ctx;
}


