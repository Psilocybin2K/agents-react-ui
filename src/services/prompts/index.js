// Prompts service: CRUD operations backed by localStorage adapter
/**
 * @typedef {import('../../types').Prompt} Prompt
 */
import { getAll, setAll, seedIfEmpty } from './storage';

const DEFAULT_SEED_FLAG = '1';

function nowMs() {
  return Date.now();
}

function generateId() {
  // Simple RFC4122-ish id; sufficient for client-only persistence
  return 'prm_' + Math.random().toString(36).slice(2, 10) + '_' + nowMs().toString(36);
}

function clampString(input, max) {
  if (typeof input !== 'string') return '';
  if (max && input.length > max) return input.slice(0, max);
  return input;
}

// Defensive limits mirrored from validation plan
const LIMITS = {
  name: 100,
  description: 500,
  content: 5000,
  tag: 50,
};

const MAX_TAGS = 10;
const MAX_VARIABLES = 50;

/**
 * Optionally seed default prompts when storage is empty.
 * Controlled by env flags: VITE_SEED_PROMPTS or REACT_APP_SEED_PROMPTS (truthy to enable).
 */
export function maybeSeed() {
  const env = (typeof process !== 'undefined' && process.env) ? process.env : {};
  const flag = env.VITE_SEED_PROMPTS ?? env.REACT_APP_SEED_PROMPTS ?? DEFAULT_SEED_FLAG;
  const enabled = String(flag) !== '0' && String(flag).toLowerCase() !== 'false';
  if (!enabled) return false;
  const ts = nowMs();
  const seed = [
    {
      id: generateId(),
      name: 'General Purpose Prompt',
      description: 'Reusable generic prompt scaffold for tasks',
      content: 'You are a helpful assistant. Use clear, concise answers.',
      variables: [],
      tags: ['general'],
      createdAt: ts,
      updatedAt: ts,
    },
  ];
  return seedIfEmpty(seed);
}

/**
 * List all prompts.
 * @returns {Promise<Prompt[]>}
 */
export async function listPrompts() {
  // Ensure seed once per session if applicable
  maybeSeed();
  const prompts = getAll();
  // Sort by updatedAt desc for stable ordering
  return prompts.sort((a, b) => (b.updatedAt || 0) - (a.updatedAt || 0));
}

/**
 * Create a new prompt
 * @param {{ name: string; content: string; description?: string; variables?: Array<{ name: string; description?: string; required?: boolean; example?: string }>; tags?: string[]; }} data
 * @returns {Promise<Prompt>}
 */
export async function createPrompt(data) {
  const name = clampString(String(data?.name || '').trim(), LIMITS.name);
  const content = clampString(String(data?.content || '').trim(), LIMITS.content);
  if (!name) throw new Error('Name is required');
  if (!content) throw new Error('Content is required');

  const variablesInput = Array.isArray(data?.variables) ? data.variables.slice(0, MAX_VARIABLES) : [];
  const variables = variablesInput
    .map(v => ({
      name: clampString(String(v?.name || '').trim(), LIMITS.tag),
      description: clampString(v?.description || '', 200),
      required: Boolean(v?.required),
      example: clampString(v?.example || '', 500),
    }))
    .filter(v => v.name);

  const tagsInput = Array.isArray(data?.tags) ? data.tags.slice(0, MAX_TAGS) : [];
  const tags = tagsInput
    .map(t => clampString(String(t || '').trim(), LIMITS.tag))
    .filter(Boolean);

  const now = nowMs();
  const prompt = {
    id: generateId(),
    name,
    description: clampString(data?.description || '', LIMITS.description),
    content,
    variables,
    tags,
    createdAt: now,
    updatedAt: now,
  };
  const prompts = getAll();
  prompts.push(prompt);
  setAll(prompts);
  return prompt;
}

/**
 * Update an existing prompt by id
 * @param {string} id
 * @param {Partial<Prompt>} updates
 * @returns {Promise<Prompt>}
 */
export async function updatePrompt(id, updates) {
  const prompts = getAll();
  const idx = prompts.findIndex(p => p.id === id);
  if (idx === -1) throw new Error('Prompt not found');
  const current = prompts[idx];

  let nextVariables = current.variables || [];
  if (updates && 'variables' in updates) {
    const input = Array.isArray(updates.variables) ? updates.variables.slice(0, MAX_VARIABLES) : [];
    nextVariables = input
      .map(v => ({
        name: clampString(String(v?.name || '').trim(), LIMITS.tag),
        description: clampString(v?.description || '', 200),
        required: Boolean(v?.required),
        example: clampString(v?.example || '', 500),
      }))
      .filter(v => v.name);
  }

  let nextTags = current.tags || [];
  if (updates && 'tags' in updates) {
    const tagsInput = Array.isArray(updates.tags) ? updates.tags.slice(0, MAX_TAGS) : [];
    nextTags = tagsInput
      .map(t => clampString(String(t || '').trim(), LIMITS.tag))
      .filter(Boolean);
  }

  const next = {
    ...current,
    ...(updates?.name !== undefined ? { name: clampString(String(updates.name).trim(), LIMITS.name) } : {}),
    ...(updates?.description !== undefined ? { description: clampString(String(updates.description), LIMITS.description) } : {}),
    ...(updates?.content !== undefined ? { content: clampString(String(updates.content).trim(), LIMITS.content) } : {}),
    ...(updates?.variables !== undefined ? { variables: nextVariables } : {}),
    ...(updates?.tags !== undefined ? { tags: nextTags } : {}),
    updatedAt: nowMs(),
  };
  if (!next.name) throw new Error('Name is required');
  if (!next.content) throw new Error('Content is required');
  prompts[idx] = next;
  setAll(prompts);
  return next;
}

/**
 * Delete a prompt by id
 * @param {string} id
 * @returns {Promise<void>}
 */
export async function deletePrompt(id) {
  const prompts = getAll();
  const next = prompts.filter(p => p.id !== id);
  setAll(next);
}


