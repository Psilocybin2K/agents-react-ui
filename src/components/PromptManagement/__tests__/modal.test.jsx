import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import PromptManagementModal from '..';
import { PromptsProvider } from '../../../services/prompts/store';

function renderModal(props = {}) {
  return render(
    <PromptsProvider>
      <PromptManagementModal isOpen onClose={() => {}} {...props} />
    </PromptsProvider>
  );
}

test('renders and validates required fields', async () => {
  renderModal();
  expect(screen.getByText('Manage Prompts')).toBeInTheDocument();
  const save = screen.getByRole('button', { name: /save/i });
  expect(save).toBeDisabled();
});

test('can create then delete a prompt', async () => {
  renderModal();
  fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'Test Prompt' } });
  fireEvent.change(screen.getByLabelText(/content/i), { target: { value: 'Hello' } });

  const save = screen.getByRole('button', { name: /save/i });
  fireEvent.click(save);

  await waitFor(() => expect(screen.getByText('Test Prompt')).toBeInTheDocument());

  // select the item in list and delete
  fireEvent.click(screen.getByRole('button', { name: /^Delete$/i }));
  const deleteButtons = await screen.findAllByRole('button', { name: /^Delete$/i });
  fireEvent.click(deleteButtons[deleteButtons.length - 1]);
});

test('basic render shows title and disabled Save', () => {
  render(
    <PromptsProvider>
      <PromptManagementModal isOpen onClose={() => {}} />
    </PromptsProvider>
  );
  expect(screen.getByText('Manage Prompts')).toBeInTheDocument();
  expect(screen.getByRole('button', { name: /save/i })).toBeDisabled();
});


