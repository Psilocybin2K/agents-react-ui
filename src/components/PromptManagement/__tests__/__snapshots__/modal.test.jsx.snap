// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: initial render structure 1`] = `
<div
  modaltype="modal"
  open=""
>
  <div
    aria-label="Prompt management dialog"
  >
    <div>
      <div>
        Manage Prompts
      </div>
      <div>
        <div>
          <section
            aria-label="Prompt list"
          >
            <div>
              <div>
                <input
                  placeholder="Search prompts"
                  value=""
                />
                <button
                  appearance="primary"
                  icon="[object Object]"
                >
                  New
                </button>
              </div>
              <div
                role="list"
              >
                <div
                  size="200"
                  style="opacity: 0.7;"
                >
                  No prompts yet
                </div>
              </div>
            </div>
          </section>
          <div />
          <section
            aria-label="Prompt form"
          >
            <form>
              <div>
                <div
                  label="Name"
                  required=""
                >
                  <input
                    disabled=""
                    maxlength="100"
                    value=""
                  />
                </div>
                <div
                  label="Description"
                >
                  <input
                    disabled=""
                    maxlength="500"
                    value=""
                  />
                </div>
                <div
                  hint="The prompt body/template"
                  label="Content"
                  required=""
                >
                  <textarea
                    disabled=""
                    maxlength="5000"
                  />
                </div>
                <div>
                  <div
                    hint="Name-only editor (skeleton)"
                    label="Variables"
                  >
                    <div>
                      <button
                        appearance="primary"
                        disabled=""
                      >
                        Add variable
                      </button>
                    </div>
                  </div>
                </div>
                <div>
                  <div
                    label="Tags"
                  >
                    <div />
                    <div
                      style="margin-top: 8px; display: flex; gap: 8px;"
                    >
                      <input
                        disabled=""
                        placeholder="Add tag"
                      />
                      <div
                        size="200"
                        style="opacity: 0.7;"
                      >
                        Press Enter to add
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </form>
          </section>
        </div>
      </div>
      <div>
        <div>
          <div>
            <div
              size="200"
              style="opacity: 0.7;"
            >
              Loadingâ€¦
            </div>
          </div>
          <div>
            <button
              appearance="secondary"
            >
              Cancel
            </button>
            <button
              appearance="primary"
              disabled=""
            >
              Save
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
